### ISSUE TITLE
How to control the behavior of enb?

### ISSUE THREAD

**Issue by JC1O24** [cncf/cnf-testbed#370]
For example, I do not want to start enb when the container started, what can I do?
I try to remove this line:
https://github.com/cncf/cnf-testbed/blob/master/examples/use_case/gogtp-k8s/k8s_bridged_networks/gogtp/templates/deployment.yaml#L19
But no luck, the container of enb pod started then become failed, any ideas?

**Comment by michaelspedersen** [comment]
Hi @JC1O24. Instead of removing the line, try replacing it with `tail -f /dev/null` or another "simple" task that will keep the container occupied

**Comment by JoeLeeWu** [comment]
> Hi @JC1O24. Instead of removing the line, try replacing it with `tail -f /dev/null` or another "simple" task that will keep the container occupied

tried. It works. Thank you so much!

**Comment by JC1O24** [comment]
Hi, @michaelspedersen. I manually start enb process by `/usr/local/bin/enb -config /root/enb.yml`. 
Then I kill it by pid running `kill -9 {pid_of enb}`. After around 3 times of starting and stopping, an error occurs like `[eNB] 2020/12/01 09:31:59 FATAL: failed to add device: gtp-enb: file exists`. Then I can not start the enb process again. Any ideas? Thanks!

**Comment by michaelspedersen** [comment]
@JC1O24 If you are running in a container I would try redeploying it. It looks like the issue is ENB trying to set up a GTP endpoint and failing to do so. Alternatively you will need to mess around with netlink, which can be a bit difficult directly through the terminal.

**Comment by JoeLeeWu** [comment]
> @JC1O24 If you are running in a container I would try redeploying it. It looks like the issue is ENB trying to set up a GTP endpoint and failing to do so. Alternatively you will need to mess around with netlink, which can be a bit difficult directly through the terminal.

In k8s pod

**Comment by michaelspedersen** [comment]
Yeah then try redeploying the pod. If that doesn't help, and you don't want to mess around with netlink, you can try rebooting the node

**Comment by JoeLeeWu** [comment]
> Yeah then try redeploying the pod. If that doesn't help, and you don't want to mess around with netlink, you can try rebooting the node

Tried. But this is not what I want. May need further studying

**Comment by michaelspedersen** [comment]
> > Yeah then try redeploying the pod. If that doesn't help, and you don't want to mess around with netlink, you can try rebooting the node
> 
> Tried. But this is not what I want. May need further studying

Otherwise going forward (while not being entirely sure if possible) it might be worth considering a different signal than SIGKILL, so the app has a change to do a graceful termination :)

**Comment by JC1O24** [comment]
Hi @michaelspedersen, when running the pgw, I found this error:
``` 
[P-GW] 2020/12/04 03:53:32 Started serving S5-C on 172.25.1.14:2123
[P-GW] 2020/12/04 03:53:32 FATAL: failed to add device: gtp-pgw: operation not supported
[P-GW] 2020/12/04 03:53:32 error reading from Conn 172.25.1.14:2123: read udp 172.25.1.14:2123: use of closed network connection
```
Also for sgw:
```
[S-GW] 2020/12/04 03:56:24 Started serving S11 on 172.22.0.13:2123
[S-GW] 2020/12/04 03:56:24 Started serving S5-C on 172.25.1.13:2123
[S-GW] 2020/12/04 03:56:24 FATAL: failed to add device: gtp-sgw-s1: operation not supported
[S-GW] 2020/12/04 03:56:24 error reading from Conn 172.22.0.13:2123: read udp 172.22.0.13:2123: use of closed network connection
[S-GW] 2020/12/04 03:56:24 error reading from Conn 172.25.1.13:2123: read udp 172.25.1.13:2123: use of closed network connection
```
But mme and enb can be started normally.
Any ideas?

**Comment by michaelspedersen** [comment]
@JC1O24 Are the bridge networks being created and available inside of the containers? It's not an error I recall ever seeing, and I'm wondering if it is due to the interfaces not being added. Regardless, I'll setup a cluster and try it out myself

**Comment by wmnsk** [comment]
FYI, I’ve seen it when the kernel GTP-U module is not loaded. So, checking the kernel version(4.12+) and the list of loaded module with `lsmod` to see if there are `gtp` and `udp_tunnel` might help.

**Comment by JoeLeeWu** [comment]
> FYI, I’ve seen it when the kernel GTP-U module is not loaded. So, checking the kernel version(4.12+) and the list of loaded module with `lsmod` to see if there are `gtp` and `udp_tunnel` might help.

Only linux version 4.12+ can work?

**Comment by electrocucaracha** [comment]
> Hi @michaelspedersen, when running the pgw, I found this error:
> 
> ```
> [P-GW] 2020/12/04 03:53:32 Started serving S5-C on 172.25.1.14:2123
> [P-GW] 2020/12/04 03:53:32 FATAL: failed to add device: gtp-pgw: operation not supported
> [P-GW] 2020/12/04 03:53:32 error reading from Conn 172.25.1.14:2123: read udp 172.25.1.14:2123: use of closed network connection
> ```
> 
> Also for sgw:
> 
> ```
> [S-GW] 2020/12/04 03:56:24 Started serving S11 on 172.22.0.13:2123
> [S-GW] 2020/12/04 03:56:24 Started serving S5-C on 172.25.1.13:2123
> [S-GW] 2020/12/04 03:56:24 FATAL: failed to add device: gtp-sgw-s1: operation not supported
> [S-GW] 2020/12/04 03:56:24 error reading from Conn 172.22.0.13:2123: read udp 172.22.0.13:2123: use of closed network connection
> [S-GW] 2020/12/04 03:56:24 error reading from Conn 172.25.1.13:2123: read udp 172.25.1.13:2123: use of closed network connection
> ```
> 
> Also for enb:
> 
> ```
> [eNB] 2020/12/04 07:03:39 Established S1-MME connection with 172.21.1.12:36412
> [eNB] 2020/12/04 07:03:39 FATAL: failed to add device: gtp-enb: operation not supported
> ```
> 
> Any ideas?

@JoeLeeWu  In my case, I have a [script](https://github.com/electrocucaracha/gw-tester/blob/master/k8s/enb.yml#L68-L118) that waits until the other services are available. Using this [approach](https://github.com/electrocucaracha/gw-tester/blob/master/k8s/deploy_demo.sh#L66-L76), I can start the pods in any order.

**Comment by wmnsk** [comment]
@JoeLeeWu 
> Only kernel version 4.12+ can work?

Yes, because it depends on Linux Kernel GTP-U introduced around that version.

**Comment by JC1O24** [comment]
> > Hi @michaelspedersen, when running the pgw, I found this error:
> > ```
> > [P-GW] 2020/12/04 03:53:32 Started serving S5-C on 172.25.1.14:2123
> > [P-GW] 2020/12/04 03:53:32 FATAL: failed to add device: gtp-pgw: operation not supported
> > [P-GW] 2020/12/04 03:53:32 error reading from Conn 172.25.1.14:2123: read udp 172.25.1.14:2123: use of closed network connection
> > ```
> > 
> > 
> > Also for sgw:
> > ```
> > [S-GW] 2020/12/04 03:56:24 Started serving S11 on 172.22.0.13:2123
> > [S-GW] 2020/12/04 03:56:24 Started serving S5-C on 172.25.1.13:2123
> > [S-GW] 2020/12/04 03:56:24 FATAL: failed to add device: gtp-sgw-s1: operation not supported
> > [S-GW] 2020/12/04 03:56:24 error reading from Conn 172.22.0.13:2123: read udp 172.22.0.13:2123: use of closed network connection
> > [S-GW] 2020/12/04 03:56:24 error reading from Conn 172.25.1.13:2123: read udp 172.25.1.13:2123: use of closed network connection
> > ```
> > 
> > 
> > Also for enb:
> > ```
> > [eNB] 2020/12/04 07:03:39 Established S1-MME connection with 172.21.1.12:36412
> > [eNB] 2020/12/04 07:03:39 FATAL: failed to add device: gtp-enb: operation not supported
> > ```
> > 
> > 
> > Any ideas?
> 
> @JoeLeeWu In my case, I have a [script](https://github.com/electrocucaracha/gw-tester/blob/master/k8s/enb.yml#L68-L118) that waits until the other services are available. Using this [approach](https://github.com/electrocucaracha/gw-tester/blob/master/k8s/deploy_demo.sh#L66-L76), I can start the pods in any order.

After my testing with Linux version which is 4.12+, the issue is gone. Thanks. But I am still finding a graceful solution for `start` and `stop` enb

**Comment by JC1O24** [comment]
> @JoeLeeWu
> 
> > Only kernel version 4.12+ can work?
> 
> Yes, because it depends on Linux Kernel GTP-U introduced around that version.

Thanks. And any ideas for `start` and `stop` enb? As I mentioned it here: https://github.com/cncf/cnf-testbed/issues/370#issuecomment-736350730

**Comment by michaelspedersen** [comment]
Thanks for the pointer @wmnsk, greatly appreciated 👍 

My local testbed is running kernel 5.4, and no issues seen there. Good that this also solved it for @JC1O24.

**Comment by JC1O24** [comment]
@michaelspedersen Hi, I try to combine enb, mme, sgw into one contanier called ems within one pod, I attached the networks that enb, mme, sgw needed. And I also added the ip addresses they need. When I start enb, the tunnels for subscribers are created succcessfully, but can not send u plane from enb to ext. 
The error is like: ```[eNB] 2020/12/17 08:00:42 WARN: failed to GET http://10.0.1.201/: Get "http://10.0.1.201/": context deadline exceeded (Client.Timeout exceeded while awaiting headers)```
Any ideas?

**Comment by michaelspedersen** [comment]
@JC1O24 From looking over this it seems that the communication goes all the way to the pgw. Do you have any logs available for that, and have you made sure that the EXT http server is running (which has the 10.0.1.201 endpoint)?

**Comment by JoeLeeWu** [comment]
> @JC1O24 From looking over this it seems that the communication goes all the way to the pgw. Do you have any logs available for that, and have you made sure that the EXT http server is running (which has the 10.0.1.201 endpoint)?

EXT is running. If I split enb, mme, sgw into 3 pods, then they work well. I want to use network namespace to isolate their communication to simulate the situation of being 3 pods. But I do not what is the better to do that.

**Comment by JoeLeeWu** [comment]
I try combine mme and sgw into one pod then leave enb in another pod, then they work well. But this is not what I want.

**Comment by michaelspedersen** [comment]
> I try combine mme and sgw into one pod then leave enb in another pod, then they work well. But this is not what I want.

For namespaces, you can try looking at the example here: https://github.com/cncf/cnf-testbed/blob/master/examples/use_case/nsmcon-ext-pf/README.md#testing-the-nsmcon-external-packet-filtering-example

Off the top of my head I can't see how this could be done, but there is probably a way.

**Comment by JoeLeeWu** [comment]
> > I try combine mme and sgw into one pod then leave enb in another pod, then they work well. But this is not what I want.
> 
> 
> 
> For namespaces, you can try looking at the example here: https://github.com/cncf/cnf-testbed/blob/master/examples/use_case/nsmcon-ext-pf/README.md#testing-the-nsmcon-external-packet-filtering-example
> 
> 
> 
> Off the top of my head I can't see how this could be done, but there is probably a way.

If I split enb, mme, sgw into 3 namespaces, how the program know to find the specific ip of enb, mme or sgw? The topology may be changed completely？
